File: ios/ExpoWeChatModuleTypes.swift
//
//  ExpoWeChatModuleTypes.swift
//  ExpoWechat
//
//  Created by Aron on 2025/6/17.
//

import Foundation
import ExpoModulesCore
import WechatOpenSDK


/**
 * 微信扫码登录所需参数。
 */
struct AuthByQROptions : Record {
    @Field
    var appId: String = ""

    @Field
    var appSecret: String = ""

    @Field
    var scope: String = "snsapi_userinfo"

    @Field
    var schemeData: String?
}

/**
 * 图片分享对象
 */
struct ShareImageOptions : Record {
    @Field
    var base64OrImageUri: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var thumbBase64OrImageUri: String?

    @Field
    var imageDataHash: String?

    @Field
    var miniProgramId: String?

    @Field
    var miniProgramPath: String?
}

/**
 * 音乐分享对象
 */
struct ShareMusicOptions : Record {
    @Field
    var musicWebpageUrl: String = ""

    @Field
    var musicFileUri: String = ""

    @Field
    var singerName: String = ""

    @Field
    var duration: Int = 0

    @Field
    var scene: String = "timeline"

    @Field
    var songLyric: String?

    @Field
    var hdAlbumThumbBase64OrImageUri: String?

    @Field
    var hdAlbumThumbFileHash: String?

    @Field
    var albumName: String?

    @Field
    var musicGenre: String?

    @Field
    var issueDate: UInt64?

    @Field
    var identification: String?

    @Field
    var title: String?

    @Field
    var description: String?

    @Field
    var extraMessage: String?

    @Field
    var thumbBase64OrImageUri: String?
}

/*
* 视频分享对象
*/
struct ShareVideoOptions : Record {
    @Field
    var videoUri: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var lowQualityVideoUri: String?

    @Field
    var thumbBase64OrImageUri: String?

    @Field
    var title: String?

    @Field
    var description: String?
}

/*
* 网页分享对象
*/
struct ShareWebpageOptions : Record {
    @Field
    var url: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var extraInfo: String?

    @Field
    var canvasPageXml: String?

    @Field
    var title: String = ""

    @Field
    var description: String = ""

    @Field
    var thumbBase64OrImageUri: String?
}

/*
* 小程序分享对象
*/
struct ShareMiniProgramOptions : Record {
    @Field
    var webpageUrl: String?

    @Field
    var id: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var path: String?

    @Field
    var type: String = "release"

    @Field
    var withShareTicket: Bool?

    @Field
    var title: String?

    @Field
    var description: String?

    @Field
    var thumbBase64OrImageUri: String?

    @Field
    var disableForward: Bool?

    @Field
    var isUpdatableMessage: Bool?

    @Field
    var isSecretMessage: Bool?
}

/**
 * 启动小程序所需参数
 */
struct LaunchMiniProgramOptions : Record {
    @Field
    var id: String = ""

    @Field
    var type: String = "release"

    @Field
    var path: String?

    @Field
    var extraData: String?
}

struct PayOptions : Record {
    @Field
    var partnerId: String = ""

    @Field
    var prepayId: String = ""

    @Field
    var nonceStr: String = ""

    @Field
    var timeStamp: UInt32 = 0

    @Field
    var sign: String = ""

    @Field
    var `package`: String = ""

    @Field
    var extraData: String = ""
}


enum LogLevel: String, Comparable {
    case verbose
    case debug
    case info
    case warning
    case error
    
    var weight: Int {
        switch self {
        case .verbose: return 0
        case .debug:   return 1
        case .info:    return 2
        case .warning: return 3
        case .error:   return 4
        }
    }
    
    var wxLogLevel: WXLogLevel {
        switch self {
        case .verbose, .debug, .info:
            return WXLogLevel.detail
        default: return WXLogLevel.normal
        }
    }
    
    static func < (lhs: LogLevel, rhs: LogLevel) -> Bool {
        return lhs.weight < rhs.weight
    }
    
    static func <= (lhs: LogLevel, rhs: LogLevel) -> Bool {
        return lhs.weight <= rhs.weight
    }
}

File: android/src/main/java/expo/modules/wechat/ExpoWeChatModuleTypes.kt
package expo.modules.wechat

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

/**
 * 微信扫码登录所需参数。
 */
class AuthByQROptions : Record {
    @Field
    var appId: String = ""

    @Field
    var appSecret: String = ""

    @Field
    var scope: String = "snsapi_userinfo"

    @Field
    var schemeData: String? = null
}

/**
 * 图片分享对象
 */
class ShareImageOptions : Record {
    @Field
    var base64OrImageUri: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var thumbBase64OrImageUri: String? = null

    @Field
    var imageDataHash: String? = null

    @Field
    var miniProgramId: String? = null

    @Field
    var miniProgramPath: String? = null
}

/**
 * 音乐分享对象
 */
class ShareMusicOptions : Record {
    @Field
    var musicWebpageUrl: String = ""

    @Field
    var musicFileUri: String = ""

    @Field
    var singerName: String = ""

    @Field
    var duration: Int = 0

    @Field
    var scene: String = "timeline"

    @Field
    var songLyric: String? = null

    @Field
    var hdAlbumThumbFilePath: String? = null

    @Field
    var hdAlbumThumbFileHash: String? = null

    @Field
    var albumName: String? = null

    @Field
    var musicGenre: String? = null

    @Field
    var issueDate: Long? = null

    @Field
    var identification: String? = null

    @Field
    var title: String? = null

    @Field
    var description: String? = null

    @Field
    var extraMessage: String? = null

    @Field
    var thumbBase64OrImageUri: String? = null
}

/*
* 视频分享对象
*/
class ShareVideoOptions : Record {
    @Field
    var videoUri: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var lowQualityVideoUri: String? = null

    @Field
    var thumbBase64OrImageUri: String? = null

    @Field
    var title: String? = null

    @Field
    var description: String? = null
}

/*
* 网页分享对象
*/
class ShareWebpageOptions : Record {
    @Field
    var url: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var extraInfo: String? = null

    @Field
    var canvasPageXml: String? = null

    @Field
    var title: String = ""

    @Field
    var description: String = ""

    @Field
    var thumbBase64OrImageUri: String? = null
}

/*
* 小程序分享对象
*/
class ShareMiniProgramOptions : Record {
    @Field
    var webpageUrl: String? = null

    @Field
    var id: String = ""

    @Field
    var scene: String = "timeline"

    @Field
    var path: String? = null

    @Field
    var type: String = "release"

    @Field
    var withShareTicket: Boolean? = null

    @Field
    var title: String? = null

    @Field
    var description: String? = null

    @Field
    var thumbBase64OrImageUri: String? = null

    @Field
    var disableForward: Boolean? = null

    @Field
    var isUpdatableMessage: Boolean? = null

    @Field
    var isSecretMessage: Boolean? = null
}

/**
 * 启动小程序所需参数
 */
class LaunchMiniProgramOptions : Record {
    @Field
    var id: String = ""

    @Field
    var type: String = "release"

    @Field
    var path: String? = null

    @Field
    var extraData: String? = null
}

class PayOptions : Record {
    @Field
    var partnerId: String = ""

    @Field
    var prepayId: String = ""

    @Field
    var nonceStr: String = ""

    @Field
    var timeStamp: Int = 0

    @Field
    var sign: String = ""

    @Field
    var `package`: String = ""

    @Field
    var extraData: String = ""
}

enum class LogLevel(val weight: Int) {
    VERBOSE(0),
    DEBUG(1),
    INFO(2),
    WARNING(3),
    ERROR(4);

    companion object {
        fun fromString(level: String): LogLevel {
            return when (level.lowercase()) {
                "verbose" -> VERBOSE
                "debug" -> DEBUG
                "info" -> INFO
                "warning" -> WARNING
                else -> ERROR
            }
        }
    }
}